src/ngx_http_lua_connection.c

```diff
@@ -0,0 +1,353 @@
+}
+
+
+/* Lua C API bindings */
+
+static int
+ngx_http_lua_connection_new(lua_State *L)
+{
+    ngx_connection_t **pconn;
+    ngx_socket_t       fd = luaL_checkinteger(L, 1);
+    const char        *errmsg;
+    luaL_argcheck(L, fd >= 0, 1, "invalid file descriptor");
+    pconn = lua_newuserdata(L, sizeof(*pconn));
+    *pconn = NULL;
+    luaL_getmetatable(L, NGX_HTTP_LUA_CONNECTION_KEY);
+    lua_setmetatable(L, -2);
+    if (NGX_ERROR == ngx_http_lua_connection_init(pconn, fd, &errmsg)) {
+        return luaL_error(L, errmsg);
+    }
+    (*pconn)->data = NULL;
+    return 1;
+}
+
+
+static ngx_connection_t*
+ngx_http_lua_connection_check(lua_State *L, int idx)
+{
+    ngx_connection_t **pconn = luaL_checkudata(L, idx, NGX_HTTP_LUA_CONNECTION_KEY);
+    luaL_argcheck(L, *pconn != NULL, idx, "ngx_connection_t has been freed");
+    return *pconn;
+}
+
+
+static int
+ngx_http_lua_connection_gc(lua_State *L)
+{
+    ngx_connection_t **pconn = luaL_checkudata(L, 1, NGX_HTTP_LUA_CONNECTION_KEY);
+    if (NULL != *pconn) {
+        ngx_http_lua_connection_release(*pconn);
+        *pconn = NULL;
+    }
+    return 0;
+}
+
+
+static int
+ngx_http_lua_connection_tostring(lua_State *L)
+{
+    ngx_connection_t *conn = ngx_http_lua_connection_check(L, 1);
+    lua_pushfstring(L, "ngx_connection*: %p", conn);
+    return 1;
+}
+
+
+static int
+ngx_http_lua_connection_wait(lua_State *L)
+{
+    ngx_connection_t   *conn = ngx_http_lua_connection_check(L, 1);
+    int                 poll_mask;
+    double              timeout;
+    const char         *events, *err;
+    ngx_http_request_t *r;
+
+    switch(lua_type(L, 2)) {
+    case LUA_TSTRING:
+        events = lua_tostring(L, 2);
+        poll_mask = 0;
+        while (*events) {
+            if (*events == 'r') {
+                poll_mask |= POLLIN;
+            } else if (*events == 'w') {
+                poll_mask |= POLLOUT;
+            }
+            events++;
+        }
+        break;
+    case LUA_TNUMBER:
+        poll_mask = lua_tointeger(L, 2);
+        if (!(poll_mask & ~(POLLIN|POLLOUT))) {
+            break;
+        }
+        /* fall through on invalid poll_mask */
+    default:
+        return luaL_argerror(L, 2, "expected bitwise 'or' of POLLIN|POLLOUT or characters from set 'rw'");
+    }
+
+    timeout = luaL_optnumber(L, 3, HUGE_VAL); /* default to infinite timeout */
+    if (!poll_mask && timeout == HUGE_VAL) {
+        return luaL_error(L, "must provide a valid events mask or finite timeout");
+    }
+    if ((r = ngx_http_lua_get_req(L)) == NULL) {
+        return luaL_error(L, "no request found");
+    }
+
+    if (NGX_ERROR == ngx_http_lua_connection_prep(r, conn, poll_mask,
+        (timeout < 0)? 0 : (timeout == HUGE_VAL)? ((ngx_msec_t)-1) : timeout*1000, &err)) {
+        return luaL_error(L, err);
+    }
+
+    return lua_yield(L, 0);
+}
+
+
+void
+ngx_http_lua_inject_connection(lua_State *L)
+{
+    luaL_newmetatable(L, NGX_HTTP_LUA_CONNECTION_KEY);
```

Comment:
I suggest hooking the metatable onto the `ngx.connection` module table. So the user can speed up her code by writing something like this:

```lua
local ngx_connection = require "ngx.connection"
local wait_read = ngx_connection.wait_read
...
local ok, err = wait_read(c)
```